google-cloud-storage/src/main/java/com/google/cloud/storage/ThroughputMovingWindow.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import com.google.common.base.MoreObjects;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Comparator;
+import java.util.PriorityQueue;
+
+/**
+ * A simple moving window implementation which will keep a {@code window}s worth of Throughput
+ * values and allow querying for the aggregate avg over that time window.
+ */
+final class ThroughputMovingWindow {
+
+  private final Duration window;
+
+  private final PriorityQueue<Entry> values;
+
+  private ThroughputMovingWindow(Duration window) {
+    this.window = window;
+    this.values = new PriorityQueue<>(Entry.COMP);
+  }
+
+  void add(Instant now, Throughput value) {
+    houseKeeping(now);
+    values.add(new Entry(now, value));
+  }
+
+  Throughput avg(Instant now) {
+    houseKeeping(now);
+    return values.stream()
+        .map(Entry::getValue)
+        .reduce(
+            Throughput.zero(),
+            (tp1, tp2) -> Throughput.of(tp1.getNumBytes() + tp2.getNumBytes(), window));
+  }
+
+  private void houseKeeping(Instant now) {
```

Comment:
nit: houseKeeping is a little confusing, I had to read through the code a few times to figure out its purpose. maybe shiftWindow or something...

Reply:
We can come up with a different name. The idea of this method is to trim the entries we're tracking to those that fit within the time window relative to `now`.

The advantage of this approach is we don't need a background thread to do any cleanup or bounding, each operation done with the Moving Window implicitly does that.

Reply:
makes sense, it would be beneficial to either add a comment explaining this or renaming of the method (or both).

Reply:
renamed the method to `removeExpiredEntries`